runtime/storage/firebase-storage.js

```diff
@@ -143,149 +144,463 @@ class FirebaseStorageProvider extends StorageProviderBase {
+  _remoteValueChanged(dataSnapshot) {
+    if (this._localModified) {
+      return;
+    }
+    if (dataSnapshot.version <= this._version) {
+      return;
+    }
+
+    let data = dataSnapshot.val();
+    this._value = data.value;
+    this._version = data.version;
+
+    this._resolveInitialized();
+    this._fire('change', {data: data.value, version: this._version});
+  }
+
+  get _hasLocalChanges() {
+    return this._localModified;
+  }
+  
+  async _persistChangesImpl() {
+    assert(this._localModified);
+    // Guard the specific version that we're writing. If we receive another
+    // local mutation, these versions will be different when the transaction
+    // completes indicating that we need to continue the process of sending
+    // local modifications.
+    let version = this._version;
+    let value = this._value;
+    let result = await realTransaction(this.reference, data => {
+      assert(this._version >= version);
+      return {
+        version: Math.max(data.version + 1, version),
+        value: value,
+      };
+    });
+    assert(result.committed, 'uncommited transaction (offline?) not supported yet');
+    let data = result.snapshot.val();
+    assert(data.version >= this._version);
+    if (this._version != version) {
+      // A new local modification happened while we were writing the previous one.
+      return this._persistChangesImpl();
+    }
+
+    this._localModified = false;
+    this._version = data.version;
+    this._value = data.value;
   }
 
   async get() {
-    return this.dataSnapshot.val().data;
+    await this._initialized;
+    return this._value;
   }
 
-  async getWithVersion() {
-    if (this.dataSnapshot == undefined) {
-      return new Promise((resolve, reject) => {
-        this._pendingGets.push(resolve);
-      });
+  async set(value, originatorId=null, barrier=null) {
+    if (this._version == null) {
+      assert(!this._localModified);
+      // If the first modification happens before init, this becomes
+      // init. We pick the initial version which will be updaetd by the
+      // transaction in _persistChanges.
+      this._version = 0;
+      this._resolveInitialized();
+    } else {
+      if (JSON.stringify(this._value) == JSON.stringify(value))
+         return;
+      this._version++;
     }
-    return this.dataSnapshot.val();
+    this._value = value;
+    this._fire('change', {data: this._value, version: this._version, originatorId, barrier});
+    await this._persistChanges();
   }
 
-  async _setWithVersion(data, version) {
-    await realTransaction(this.reference, _ => ({data, version}));
+  async clear(originatorId=null, barrier=null) {
+    return this.set(null, originatorId, barrier);
   }
 
-  async set(value) {
-    return realTransaction(this.reference, data => {
-      if (JSON.stringify(data.data) == JSON.stringify(value))
-        return data;
-      return {data: value, version: data.version + 1};
-    });
+  async cloneFrom(handle) {
```

Comment:
Move this to StorageProviderBase? I think it's the same for all implementations.